### ISSUE TITLE
The `mysql2` gem does not compile on Fedora 27/28.

### ISSUE THREAD

**Issue by pketienne** [psu-libraries/scholarsphere-3#1322]
When attempting to perform a bundle install for scholarsphere on operating systems Fedora 27 and 28, compilation of the extensions for the `mysql2` gem fail to build. A full stacktrace will be posted in a follow-up comment. In order to circumvent this issue, the following steps can be used:

1. Change line 7 of the Gemfile from `gem 'mysql2', '~> 0.3.17' unless ENV['CI']` to `gem 'mysql2', '~> 0.4.7' unless ENV['CI']`.
1. Change line 500 of the Gemfile.lock from `mysql2 (0.3.21)` to `mysql2 (0.4.10)`.
1. Change line 923 of the Gemfile.lock from `mysql2 (~> 0.3.17)` to `mysql2 (~> 0.4.10)`.
1. `RAILS_ENV='production' bundle install`

In short, Fedora 27/28 cannot compile extensions for the `mysql2` gem for versions below 0.4.7.

**Comment by pketienne** [comment]
Installing mysql2 0.3.21 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /home/pke3/.rbenv/versions/2.3.3/gemsets/scholarsphere/gems/mysql2-0.3.21/ext/mysql2
/home/pke3/.rbenv/versions/2.3.3/bin/ruby -r ./siteconf20180719-32161-1xnxj6s.rb extconf.rb
checking for ruby/thread.h... yes
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
checking for rb_thread_blocking_region()... no
checking for rb_wait_for_single_fd()... yes
checking for rb_hash_dup()... yes
checking for rb_intern3()... yes
-----
Using mysql_config at /usr/bin/mysql_config
-----
checking for mysql.h... yes
checking for errmsg.h... yes
checking for mysqld_error.h... yes
-----
Setting libpath to /usr/lib64/
-----
creating Makefile

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /home/pke3/.rbenv/versions/2.3.3/gemsets/scholarsphere/extensions/x86_64-linux/2.3.0-static/mysql2-0.3.21/mkmf.log

current directory: /home/pke3/.rbenv/versions/2.3.3/gemsets/scholarsphere/gems/mysql2-0.3.21/ext/mysql2
make "DESTDIR=" clean

current directory: /home/pke3/.rbenv/versions/2.3.3/gemsets/scholarsphere/gems/mysql2-0.3.21/ext/mysql2
make "DESTDIR="
compiling infile.c
In file included from ./mysql2_ext.h:18,
                 from infile.c:1:
/usr/include/mysql/mysql_com.h:3:2: warning: #warning This file should not be included by clients, include only <mysql.h> [-Wcpp]
 #warning This file should not be included by clients, include only <mysql.h>
  ^~~~~~~
compiling result.c
In file included from ./mysql2_ext.h:18,
                 from result.c:1:
/usr/include/mysql/mysql_com.h:3:2: warning: #warning This file should not be included by clients, include only <mysql.h> [-Wcpp]
 #warning This file should not be included by clients, include only <mysql.h>
  ^~~~~~~
result.c: In function ‘msec_char_to_uint’:
result.c:189:17: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
   for (i = 0; i < (len - 1); i++) {
                 ^
result.c: In function ‘rb_mysql_result_fetch_row’:
result.c:259:14: warning: this statement may fall through [-Wimplicit-fallthrough=]
           if (castBool && fields[i].length == 1) {
              ^
result.c:263:9: note: here
         case MYSQL_TYPE_SHORT:      /* SMALLINT field */
         ^~~~
compiling client.c
In file included from ./mysql2_ext.h:18,
                 from client.c:1:
/usr/include/mysql/mysql_com.h:3:2: warning: #warning This file should not be included by clients, include only <mysql.h> [-Wcpp]
 #warning This file should not be included by clients, include only <mysql.h>
  ^~~~~~~
In file included from /home/pke3/.rbenv/versions/2.3.3/include/ruby-2.3.0/ruby/ruby.h:1992,
                 from /home/pke3/.rbenv/versions/2.3.3/include/ruby-2.3.0/ruby.h:33,
                 from ./mysql2_ext.h:8,
                 from client.c:1:
client.c: In function ‘rb_mysql_client_info’:
client.c:60:30: error: ‘MYSQL_SERVER_VERSION’ undeclared (first use in this function); did you mean ‘MYSQL_LINK_VERSION’?
   #define MYSQL_LINK_VERSION MYSQL_SERVER_VERSION
                              ^~~~~~~~~~~~~~~~~~~~
/home/pke3/.rbenv/versions/2.3.3/include/ruby-2.3.0/ruby/intern.h:806:27: note: in definition of macro ‘rb_str_new_cstr’
     (__builtin_constant_p(str)) ?  \
                           ^~~
client.c:851:32: note: in expansion of macro ‘MYSQL_LINK_VERSION’
   header_version = rb_str_new2(MYSQL_LINK_VERSION);
                                ^~~~~~~~~~~~~~~~~~
client.c:60:30: note: each undeclared identifier is reported only once for each function it appears in
   #define MYSQL_LINK_VERSION MYSQL_SERVER_VERSION
                              ^~~~~~~~~~~~~~~~~~~~
/home/pke3/.rbenv/versions/2.3.3/include/ruby-2.3.0/ruby/intern.h:806:27: note: in definition of macro ‘rb_str_new_cstr’
     (__builtin_constant_p(str)) ?  \
                           ^~~
client.c:851:32: note: in expansion of macro ‘MYSQL_LINK_VERSION’
   header_version = rb_str_new2(MYSQL_LINK_VERSION);
                                ^~~~~~~~~~~~~~~~~~
client.c: In function ‘init_mysql2_client’:
client.c:60:30: error: ‘MYSQL_SERVER_VERSION’ undeclared (first use in this function); did you mean ‘MYSQL_LINK_VERSION’?
   #define MYSQL_LINK_VERSION MYSQL_SERVER_VERSION
                              ^~~~~~~~~~~~~~~~~~~~
client.c:1219:30: note: in expansion of macro ‘MYSQL_LINK_VERSION’
   for (i = 0; lib[i] != 0 && MYSQL_LINK_VERSION[i] != 0; i++) {
                              ^~~~~~~~~~~~~~~~~~
make: *** [Makefile:239: client.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/pke3/.rbenv/versions/2.3.3/gemsets/scholarsphere/gems/mysql2-0.3.21 for inspection.
Results logged to /home/pke3/.rbenv/versions/2.3.3/gemsets/scholarsphere/extensions/x86_64-linux/2.3.0-static/mysql2-0.3.21/gem_make.out

An error occurred while installing mysql2 (0.3.21), and Bundler cannot continue.
Make sure that `gem install mysql2 -v '0.3.21'` succeeds before bundling.

In Gemfile:
  mysql2

**Comment by carolyncole** [comment]
@informaticianme does this answer help? https://stackoverflow.com/questions/3608287/error-installing-mysql2-failed-to-build-gem-native-extension
